### DISCUSSION TITLE
FAQ

### DISCUSSION THREAD

**Discussion by vladmandic** [vladmandic/sdnext#1011] [Q&A]
I'd like to start an **FAQ** section and eventually bring it into `README` itself.  

So this is an ask to contribute:  **One Q + A per post**
(and lets not have a long conversations in this thread, this is all about gathering material for FAQ)

I'll start...

Q: Where is "PNG Info" tab
A: Full functionality is integrated into "Process Image" tabs and there is no need for a separate tab

Q: Where are command line flags like `--xformers`
A: Most of command line flags have been moved to UI -> Settings  
  For list of available command line flags, simply start `webui --help`

Q: How can I get more information about what is going on with my app?
A: Start it with `--debug` command line flag and also check `setup.log`

**Comment by brknsoul** [comment]
Q: How do I add command-line options to `webui.bat` in Windows?
A:
- Right-click on `webui.bat` and click "Create Shortcut"
- Right-click the newly created shortcut and click "Properties"
- At the end of the "Target" field, type a space, then add your options there.

It should look similar to this; 
![image](https://github.com/vladmandic/automatic/assets/1969381/663645af-33db-4543-ab3b-2e4cd6e5391b)

- Then click "OK"
- Next time you want to launch SD.Next, use this shortcut.

**Reply by IrishAI** [reply]
For this to work you mustn't remove the quotation marks on the line before you add the argument.
ie, "G:\automatic\webui.bat" --meddram --autolauch

**Reply by brknsoul** [reply]
Quotation marks are only required if the path has a space in it.
This is fine: `C:\Games\AwesomeGame\AwesomeGame.exe`
This is also fine: `"C:\More Games\Stellar Game\Stellar Game.exe"`

**Comment by BinaryQuantumSoul** [comment]
Q: How do I use an AMD GPU on Windows ?
A: You should add --use-directml command line option.

**Comment by Tillerz** [comment]
Q: How can I create large images, say 2048x2048 or higher, with a relatively low amount of VRAM?
A: Render the image as usual and then use Upscaling on the "Process Image" tab with an upscaler like SwinIR_4x which scales up an image by splitting it into smaller tiles first, keeping the memory usage low.

Q: My images are dull or just "off" colorwise.
A: Use a different VAE file. You can get a VAE selection box to the top of the page by adding ",sd_vae" to Settings / User interface / Quicksettings list: "sd_model_checkpoint,sd_vae" (and reload the UI). You then might want to find new VAE files and drop them into the models\VAE folder.

Q: I saw you updated the code, how do I get it?
A: Run `webui.bat --upgrade`. You should now see some additional lines in the output:

```
07:55:11-989641 INFO     Version: bc02d666 Tue May 16 10:52:38 2023 -0400
07:55:12-350956 INFO     Updating main repository
07:55:14-239432 INFO     Upgraded to version: 5250ba4b Tue May 16 21:20:36 2023 -0400
```

Q: Something broke. How do I get a running version of SD.next again?
A: You can either execute `webui.bat --reinstall` to reinstall all the required components or execute `webui.bat --reset` to reset your installation to the latest version from git.

Q: My image folder is exploding, how do I fix that?
A: Settings / Image options ... here you can change what is being saved and in which format. You might want to switch from default png to jpg to have smaller files and disable "Always save all generated images" and click "Save" instead on images you want to keep.

Q: I get a lot of out of memory errors, how can I fix that?
A: You can do several things:
- do only render to default size (512x512) and upscale in "Process Image" later
- Settings / Compute Settings: Set it to use FP16 instead of FP32.
- Try in Settings / Stable Diffusion which Cross-attention optimization method works best for you and make sure "SDP disable memory attention" is not checked, as this uses a lot more memory (but gives a dramatic speed improvement).

Q: I do get lots of weird output with two people merged or one with two heads etc.
A: Most models are trained on 512x512 images. If you render in different sizes (especially with different height and width), SD tries to fill that additional space, somehow.

**Reply by jimmiles** [reply]
Thank you! I thought I had plenty of VRAM, then I met Stable Diffusion.

**Comment by mark-wd** [comment]
Q: What does "Hires fix" actually do?

A: In spite of its name, the "Hires fix" is not intended to fix issues. Rather, it's a method to upscale images during the generation process itself, meaning that your image is not going to be affected by artifacts and blurriness that are the result of using an upscaler after the generation. This does not preclude the usefulness of upscalers, but rather serves as an additional tool to your workflow.

**Comment by mark-wd** [comment]
Q: What is CLIP Skip?

A: CLIP is a core element of the 1.x Stable Diffusion models (while 2.x models use OpenCLIP). In 1.x, CLIP guides the image generation process in a layered fashion, getting more specific with each layer. With the standard CLIP skip of 1, the image becomes more specific over the course of its generation by picking up characteristics of whatever path it ends up taking, with a "cat" becoming a "Siamese cat", for example. By setting a higher value of CLIP Skip, you may end up with an image assembled with less specific components, which might also be of a higher quality due to that. Results vary, but some models are specifically trained to benefit from higher CLIP Skip values.

**Comment by mark-wd** [comment]
Q: What is the best sampler to use?

A: There is no one best sampler, they all serve some purpose, but current community consensus leans towards "DPM++ 2M Karras" being a good general solution in most cases.

**Comment by Threnos** [comment]
Q: What is "Negative Guidance minimum sigma" (s_min_uncond in config)?
A: For the sake of better perfomance and sacrifising some quality, stop calculating negative after some amount of steps.
The overall look of an image is set rouphly in the 1\3 - 1\2 of steps done. After this point, only the details are being changed. Given this, we can exclude computation of nevatives to gain some performance. Of course, the quality WILL suffer.

Author personally set it to 1. This results in nevatives dropped at 30th step of 44 (which is 68%), with subjectively negligible impact on resulting image. The perfomance rises from 1.09 s/it to 1.3 it/s.

**Comment by diniamo** [comment]
Q: Why does my CLI progress bar randomly change style?/Why is UniPC the only sampler with a stylized CLI progress bar?
A: UniPC is the only sampler natively implemented in SD.Next, all the other ones come from external libraries. Modifying said libraries would add extra complications which are not worth it. More detailed answer [here](https://github.com/vladmandic/automatic/issues/935#issuecomment-1546891268)

**Comment by vashter** [comment]


**Reply by ghost** [reply]
I modify my install (just naming of files, a saving hack and I edit the color correction so mostly processing and images.py) and it caused problems in the past until I understood how git works, so maybe linking this in the answer https://tecadmin.net/force-overwrite-local-files-git-pull/#:~:text=The%20simplest%20way%20to%20force,changes%20from%20the%20remote%20repository. would be a good idea? Users can just stash the changes and re-apply them. However I never encountered a conflict and wouldnt know what to do if this happens

**Reply by TheOnlyHolyMoly** [reply]
> maybe add one about do not modify? it pops up frequently - ppl modifying webui.bat, notifications.mp3, system.css, etc.

done. thinking about it, maybe there would be sense in some kind of default report for issue filing that one could copy from system info and then paste in the issue here (system version, platform info, installed extensions, validation checksum that some essential files are unaltered). let me know if you like the thought then I would file a feature request.

**Reply by TheOnlyHolyMoly** [reply]
> I modify my install (just naming of files, a saving hack and I edit the color correction so mostly processing and images.py) and it caused problems in the past until I understood how git works, so maybe linking this in the answer https://tecadmin.net/force-overwrite-local-files-git-pull/#:~:text=The%20simplest%20way%20to%20force,changes%20from%20the%20remote%20repository. would be a good idea? Users can just stash the changes and re-apply them. However I never encountered a conflict and wouldnt know what to do if this happens

it's leading off topic here, but I'd be curious to know what mod you make on the color correction and why? can you hit me up on discord? maybe there'd be value in a feature request. I would not encourage that link though as part of Q&A as it kind of gives a sense "modify it, there's a good way back"... and tracing back modded system will just be a pain for Vlad

**Reply by vladmandic** [reply]
first 10 lines of startup log (i've cleaned it up yesterday a bit) show pretty much everything:
```log
10:38:34-363903 INFO     Starting SD.Next
10:38:34-366024 INFO     Python 3.10.6 on Linux
10:38:34-372133 INFO     Version: 0d629f3c Thu Jun 8 09:01:43 2023 -0400
10:38:34-628376 INFO     Setting environment tuning
10:38:34-634584 INFO     nVidia CUDA toolkit detected
10:38:35-302944 INFO     Torch 2.1.0.dev20230519+cu121
10:38:35-348275 INFO     Torch backend: nVidia CUDA 12.1 cuDNN 8801
10:38:35-365214 INFO     Torch detected GPU: NVIDIA GeForce RTX 3060 VRAM 12288 Arch (8, 6) Cores 28
10:38:35-380317 INFO     Enabled extensions-builtin: ['stable-diffusion-webui-images-browser', 'sd-webui-agent-scheduler', 'clip-interrogator-ext', 'ScuNET', 'stable-diffusion-webui-rembg', 'sd-extension-system-info', 'SwinIR', 'a1111-sd-webui-lycoris', 'LDSR', 'Lora', 'sd-dynamic-thresholding', 'sd-webui-controlnet', 'multidiffusion-upscaler-for-automatic1111']
10:38:35-381909 INFO     Enabled extensions: ['a1111-sd-webui-tagcomplete', 'seed_travel', 'sd-extension-aesthetic-scorer', 'adetailer', 'Stable-Diffusion-Webui-Civitai-Helper', 'stable-diffusion-webui-inspiration', 'stable-diffusion-webui-promptgen']
```
only thing missing is "has something been altered" - i can add that, can you open a feature request so i don't forget?

**Reply by TheOnlyHolyMoly** [reply]
will do

**Reply by TheOnlyHolyMoly** [reply]
is CPU and RAM a thing to add there maybe? I recall some conversation on discord where you said some issues came from insufficient cpu and bottlenecking events?

**Comment by TheOnlyHolyMoly** [comment]
Q: How do enable auto-update upon start of webui to the latest commit 
A: Run webui.bat --upgrade, no more need for manual "git pull" // Additionally, if you want to make this a default behaviour update your start parameters > https://github.com/vladmandic/automatic/discussions/1011#discussioncomment-5922369

**Comment by TheOnlyHolyMoly** [comment]
Q: I want to manually tweak my system and modify internal files (webui.bat, notifications.mp3, system.css), can you give me advice?
A: Please do NOT touch those files or any other internal files. It might do what you think it does or something completely different, In any case it will make it impossible to EVER help you if you have an issue and previously made a manual change. Use the officially documented parameters, toggles, switches. If you - in spite of this good advice - decide to modify internal system files please do NOT report issues that occur on your system. Will be impossible to reproduce.

**Reply by DrJKL** [reply]
Want to list the alternatives, like `webui-user.bat` and `user.css`?

**Reply by TheOnlyHolyMoly** [reply]
referring this topic to @vladmandic

**Reply by vladmandic** [reply]
i think it might be helpful to name some common ones. of course, user can use anything, but having examples is good.

**Comment by VStudioAI** [comment]
Q: How can I organize my ckpt files into some sort of category structure?
A: simply create well named sub folders under the models directory, move ckpts to relevant directories and restart UI. Examples might be, by type, by style etc. Photorealistic/Cartoonish/Anime etc.

NOTE: While it's true that this works it's a bit of a hinky workaround because it's difficult to read in the models drop down, the box is not always wide enough for a long path name. I think it would be a great idea if there were a variable right next to the dropdown where you could say "show me only those that are in the xyz" directory.  Anime as an example, based on your own folder designations.

**Comment by mark-wd** [comment]
Q: How do I change my SD.Next install between different disks?

A: You can simply drag and drop the root folder between disks. However, some extra steps are necessary for it to work like normal. First, delete the "venv" folder. Second, open "config.json" and change every occurence of the disk letter to the new one.

Finally, if you get warnings about "_dubious ownership_" on the prompt, you will need to tweak some permissions. **Only in case you are running your installation offline**, you can then simply do a _git config --global --add safe.directory "*"_ and you will be done.

**Comment by vujuha** [comment]
Q: How to do 4k pictures with low vram?
A: Send picture to img2img and choose Sd upscaler from script at the bottom of the ui. You should keep dimensions at 512x512 and denoising strength low, like 30-40. This will create an upscaled picture, just generating 512x512 part part so your vram usage will be like generating bunch of 512x512 pictures

Q:How to put multiple command arguments(command lines) into the webui.bat?
A:  set COMMANDLINE_ARGS=--listen
     set COMMANDLINE_ARGS=%COMMANDLINE_ARGS% --upgrade
     set COMMANDLINE_ARGS=%COMMANDLINE_ARGS% --disable-queue  
     ....

Q:How to use the webui locally from phone or another pc?
   1.Put --listen command line to Webui.bat.Save and close
   2.Open port 7860 by doing this steps below
       -Navigate to Control Panel -> System and Security -> Windows Firewall
       -Select Advanced Settings and highlight Inbound Rules in the left pane.
       -Right-click Inbound Rules and select New Rule or select New Rule from Actions on the left
       -Choose Port and click Next
       -Choose TCP and enter 7860 in Specific local ports and click Next.
       -Add the protocol (TCP)and port 7860 into the next window and click Next.
       -Select Allow the connection in the next window and hit Next.
       -Select the network types as you see fit and click Next.
       -Name the rule something meaningful and click Finish. I named it "SD PORT 7860" for example
   3. Now launch webui-user.bat as normal. It will now say Running on local URL: http://0.0.0.0:7860/. But this doesn't work directly usually so you need your IP for it to work.
   4. To get your IP, open a CMD shell (type cmd into your windows search bar) and type ipconfig. You are looking for the IPv4 Address. On the machine that you want to work from, open a browser and type http://xxx.xx.x.xx:7860/ where the XXX.XX.X.XX is the IP of the machine running the webui service.

Note: You need to save pictures with "Save as" since if you save a picture from webui it will save to the original pc.

**Reply by vladmandic** [reply]
thanks. i'd remove the last one as its really generic and `--disable-queue` is not a fix-all nor it should be used without understanding what it does.

**Comment by vujuha** [comment]
Q:How to startup webui.bat on windows start ?
A: Find the webui.bat right click it and send to desktop or create a shortcut
   -Press Windows logo key  + R, 
   -Type shell:startup, then select OK. This opens the Startup folder.
   -Copy and paste the shortcut to the app from the file location to the Startup folder.

Note: It is really useful if you use SD from another device and original one used by someone else.

**Comment by vladmandic** [comment]
> How do I ensure my UI is "private"? I see the mention of "server" a lot and it kinda bothered me

by default server runs on your pc and is accessible on your pc only
if you add `--listen` flag, its available on your local network.
if in addition to that, if you explicitly create enable reverse route on your router, it can be accessed externally
if you add `--share` flag, its available on shared gradio site

> Whats the difference betwen fresh resart of webUI and server restart from ui?

already loaded python modules can not be unloaded from memory, so if there are differences to any of those modules, warm restart may not be enough (e.g., if you disable extension, warm restart may not disable it completely as parts may still be in memory).

> Are server restarts reated to extensions?

many server settings are only applied on server startup, not related just to extensions

**Comment by rocky533** [comment]
AMD users - my card is a 580 8gb 
I had issues forever with inpaint, black boxes, white boxes, memory errors, flying donkies... wait... 
- now working up to 1024x1024 default or upscale to that with second pass option --- terrible option to do this anyway.
- inpainting fully working
- 22 second 600x800 images
=======

Many settings new here from our old interface.
First the command line args
--medvram - just use it even with more than 8 gb, it will save u trouble and does not perform much worse
--use-directml  - this is needed and MUST be used your first run, or you will get stuck in cpu mode and be forced to reinstall.
--experimental - this is what stops the webui from resetting no half and precision stuff when you reload the command prompt.

I don't need the no half and all that anymore... perhaps u don't either. The rest of the settings are in the UI.

OPTIMIZATION tab
Cross-attention optimization method  --xFormers not for us ignore it
Doggettx's and InvokeAI's - slow
Scaled-Dot-Product -- fast but unstable, mem errors like mad at just random times.
Split attention - faster than Sub-quadratic by a small amount but more memory errors.
Sub-quadratic - fast enough and uses by far the least memory and allows the largest image sizes both in render and upscale. - my option

SDP disable memory attention = makes the images come out more like when using xformers, but uses memory hard. not worth

Chunk sizes are all default for me as I am not sure what they do yet

TIP -- Sub-quadratic breaks with token merging, you get black boxes and failed images all over. However, it seems to work ok with Split attention.  Test this one for your self, as token merging and split might be your best option.

Token merging
You can enable token merging, however IMO something is broken with it and is the cause for many of the white/black images so I stick with Sub-quadratic and no merging.

COMPUTE SETTINGS tab - this is where things get different, remember the --experimental line we added, this allows us to turn things off and have them remain off on this tab, if you don't add it, changing things on this page won't stick after resetting the command window

![page](https://github.com/vladmandic/automatic/assets/140274539/1c2508fd-edae-4f5a-bad5-95d576bbc9e4)

![page2](https://github.com/vladmandic/automatic/assets/140274539/67df826b-16dd-46db-8ef4-d958811d1b2c)

I know most people are tying the same settings from older ui's, half, no half, precision.  It does not seem to be needed now, and u can use FP16 to save memory.  Please note, you're setting defaults not restrictions Precision autocast at the top still allows fp32 when the model requires it... fp32 models take much more gpu memory and are harder to render with my card.  I stick to fp16 models.

These are my settings and I get much less mem errors (still happen after long periods) and no other issues including inpaint/outpaint/upscales

**Comment by thetzomaster** [comment]
is there a way to train your own model ? if yes can you please explain how ? Thank-you :)

**Reply by vladmandic** [reply]
i'd suggest to check out https://github.com/Nerogar/OneTrainer

**Comment by brknsoul** [comment]
#### How to get SD.Next to remember settings.

- Start SD.Next.
- Adjust all your settings (eg height, width, sampler, steps, pos and/or neg prompts, etc) and open any accordions (eg batch, seed, advanced, second pass etc) you want to stay open.
- Then go to the System Tab, then to User Interface Tab (not the Settings category).
- Check that the Changed Values are what you what, then click "Set new defaults", then click "Set menu states"

![image](https://github.com/vladmandic/automatic/assets/1969381/2333ba64-d055-4123-9aa5-b0e45a25f39e)

**Reply by AndrewRainfall** [reply]
One more step: go to sdnext folder and backup the 'config.json' and 'ui-config.json' files to quickly restore your setting after reinstallation.
